electronics_lib/PowerConditioning.py

```diff
@@ -263,3 +264,136 @@ def connected_from(self, gnd: Optional[Port[VoltageLink]] = None, pwr_hi: Option
+         ),
+        'gnd': Ground(),
+      })
+
+
+
+class PmosChargerReverseProtection(PowerConditioner, KiCadSchematicBlock, Block):
+  class PmosChargerReverseProtection(PowerConditioner, KiCadSchematicBlock, Block):
+    """PMOS Charger Reverse Protection.
+
+    PMOS transistors are used in charging circuits to prevent reverse voltage from damaging the battery or the circuit.
+    They act as a switch that allows current flow only in the correct direction. Regular PMOS reverse protection may not be suitable for bidirectional circuit.
+
+    Ports:
+        - pwr_in: Power in from the battery
+        - pwr_out: Power out for the load
+        - chg_in: Charger input to charge the battery
+        - chg_out: Charging output for the battery
+        - gnd: Ground
+
+    Note:
+        - Limitation: The highest Vgs/Vds are bounded by the transistors' specifications.
+        - There is also a rare case when this circuit being disconnected when a charger is connected first. But always reverse protect.
+        - R1 and R2 are the pullup bias resistors for mp1 and mp2 PFet.
+
+    More info at: https://www.edn.com/reverse-voltage-protection-for-battery-chargers/
+    """
+
+  @init_in_parent
+  def __init__(self, r1_val: RangeLike = 100*kOhm(tol=0.01), r2_val: RangeLike = 100*kOhm(tol=0.01), rds_on: RangeLike =(0, 0.1)*Ohm):
+    super().__init__()
+
+    self.gnd = self.Port(Ground.empty(), [Common])
+    self.pwr_out = self.Port(VoltageSource.empty(),)  # Load
+    self.pwr_in = self.Port(VoltageSink.empty(), )  # For connecting battery pwr output
+    self.chg_in = self.Port(VoltageSink.empty(), )  # Charger input
+    self.chg_out = self.Port(VoltageSource.empty(),)  # Charging output to lipo
+
+    self.r1_val = self.ArgParameter(r1_val)
+    self.r2_val = self.ArgParameter(r2_val)
+    self.rds_on = self.ArgParameter(rds_on)
+
+  def contents(self):
+    super().contents()
+    self.r1 = self.Block(Resistor(resistance=self.r1_val))
+    self.r2 = self.Block(Resistor(resistance=self.r2_val))
+
+    batt_voltage = self.pwr_in.link().voltage.hull(self.chg_out.link().voltage).hull((0, 0))
+
+    # taking the max of the current for the both direction. 0 lower bound
+    batt_current = self.pwr_in.link().current_limits.hull(self.chg_out.link().current_drawn).hull((0, 0))
+    power = batt_current * batt_current * self.rds_on
+    r1_current = batt_voltage / self.r1.resistance
+
+    # Create the PMOS transistors and resistors based on the provided schematic
+    self.mp1 = self.Block(Fet.PFet(
+      drain_voltage=batt_voltage, drain_current=r1_current,
+      gate_voltage=(-batt_voltage.upper(), batt_voltage.upper()),
+      rds_on=self.rds_on,
+    ))
+    self.mp2 = self.Block(Fet.PFet(
+      drain_voltage=batt_voltage, drain_current=batt_current,
+      gate_voltage=(- batt_voltage.upper(), batt_voltage.upper()),
+      rds_on=self.rds_on,
+      power=power
+    ))
+
+    chg_in_adapter = self.Block(PassiveAdapterVoltageSink())
```

Comment:
Hm, if you just attached two ports on the same net in the schematic and use `conversions`, does it recognize that?

Reply:
If I try to do like this, it results in:
```python
self.file_path("resources", f"{self.__class__.__name__}.kicad_sch"),
      conversions={
        'pwr_in': VoltageSink(
          current_draw=batt_current
        ),
        'chg_out': VoltageSource(
          voltage_out=batt_voltage),
        'pwr_out': VoltageSource(
          voltage_out=batt_voltage),
        'gnd': Ground(),
      })
```
![image](https://github.com/BerkeleyHCI/PolymorphicBlocks/assets/49264928/b076b6af-67d0-43da-88d5-8f4f169aae1a)


```
AssertionError: attempted to reassign (adapter)mp1.gate=electronics_model.PassivePort.PassiveAdapterVoltageSink@70

AssertionError: attempted to reassign (adapter)mp1.gate=electronics_model.PassivePort.PassiveAdapterVoltageSink@70

AssertionError: attempted to reassign (adapter)mp1.gate=electronics_model.PassivePort.PassiveAdapterVoltageSink@00

ValueError: can't follow pmos.gnd from (root), have unused path elements [name: "gnd"
] from pmos
```

Reply:
keeping the current workaround